x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.test.ts

```diff
@@ -56,13 +57,55 @@ describe('perform_bulk_action_schema', () => {
       query: 'name: test',
       action: 'unknown',
     };
+    const message = retrieveValidationMessage(payload as unknown as PerformBulkActionSchema);
 
-    const decoded = performBulkActionSchema.decode(payload);
-    const checked = exactCheck(payload, decoded);
-    const message = foldLeftRight(checked);
     expect(getPaths(left(message.errors))).toEqual([
       'Invalid value "unknown" supplied to "action"',
+      'Invalid value "undefined" supplied to "updates"',
     ]);
     expect(message.schema).toEqual({});
   });
+
+  test('missing updates array is invalid when action type is update', () => {
+    const payload = {
+      query: 'name: test',
+      action: BulkAction.update,
+    };
+
+    const message = retrieveValidationMessage(payload as PerformBulkActionSchema);
+
+    expect(getPaths(left(message.errors))).toEqual([
+      'Invalid value "update" supplied to "action"',
```

Comment:
I think it happens because `updates` property is missing. So, it might be validation function treats `BulkAction.update` as invalid as it can be valid only if `updates` property is present

Reply:
Hmm, this error message looks a bit misleading to me. Does `io-ts` allow to adjust it somehow?

Reply:
++ for adjusting it if it's possible

Reply:
Doesn't look that proposed by @xcrzx [library](https://github.com/klesun/io-ts-better-union-error-reporter) helps to eliminate additional error messages:

So, on payload
```Javascript
{
    query: 'name: test',
    action: BulkAction.edit,
}
```

Our current error messages:

```Javascript
[
  'Invalid value "edit" supplied to "action"',
  'Invalid value "undefined" supplied to "edit"',
]
```

Validation errors with newly proposed library:

```
must satisfy either of
      | invalid 1 ExactType element(s)
        action: must satisfy either of
          | 'delete' expected, but 'edit' found
          | 'disable' expected, but 'edit' found
          | 'duplicate' expected, but 'edit' found
          | 'enable' expected, but 'edit' found
          | 'export' expected, but 'edit' found
      | invalid 1 ExactType element(s)
        edit: object[] is mandatory
```

Still, there is present a message regarding invalid action.

Possible solutions?

1. What if we put `edit` property as partial? 

```
export const performBulkActionSchema = t.exact(
  t.type({
    query: queryOrUndefined,
    action: bulkAction,
    [BulkAction.edit]: t.union([t.array(bulkActionEditPayload), t.undefined]),
  })
);
```
We can catch if edit is supplied with other actions(in body of route validation function) and throw error. 

Downsides: 
 - validation logic in 2 places

2. Dig deeper into union validation/error reporting, write own path reporter(??).  In that case I would suggest to handle it in a separate issue.

What are your thoughts @banderror , @spong , @xcrzx?

Reply:
This doesn't look anywhere better 🙂 

```
must satisfy either of
      | invalid 1 ExactType element(s)
        action: must satisfy either of
          | 'delete' expected, but 'edit' found
          | 'disable' expected, but 'edit' found
          | 'duplicate' expected, but 'edit' found
          | 'enable' expected, but 'edit' found
          | 'export' expected, but 'edit' found
      | invalid 1 ExactType element(s)
        edit: object[] is mandatory
```

Imho validation error messages should be clear for any user who doesn't fully understand the request/response schema and of course TS union types or any programming concepts like that. The whole `invalid 1 ExactType element(s)` line is just noise for the user.

> In that case I would suggest to handle it in a separate issue.

I agree with that. @vitaliidm could you please write up and open a ticket?

Reply:
@banderror https://github.com/elastic/kibana/issues/121860